### PR TITLE
[WPE][GTK] Remove pkg-config use from FindLibBacktrace.cmake

### PR DESCRIPTION
**Pull request by mcatanzaro**
#### 0ff5cc83058a1ec8db37ffa6784eed55eaa232b5
<pre>
[WPE][GTK] Remove pkg-config use from FindLibBacktrace.cmake
<a href="https://bugs.webkit.org/show_bug.cgi?id=265813">https://bugs.webkit.org/show_bug.cgi?id=265813</a>

Reviewed by Adrian Perez de Castro.

libbacktrace upstream does not distribute a pkg-config file. Apparently
some downstreams are doing this, but I&apos;m not sure who, because this
software is not packaged in major distros like Debian or Fedora. It&apos;s
not likely to be, because it doesn&apos;t have any releases. Let&apos;s not
attempt to use a downstream-only pkg-config file. If pkg-config support
can be upstream, that would be great.

* Source/cmake/FindLibBacktrace.cmake:

Canonical link: <a href="https://commits.webkit.org/272292@main">https://commits.webkit.org/272292@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/1abc212df75196c99810742f1a306c0bf8a9cb71

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/30482 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/9154 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/32102 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/32991 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/27572 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/11392 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/6396 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/27518 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/30790 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/7697 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/27321 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/6600 "Passed tests") | 
| | [❌ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/6746 "Too many flaky failures: imported/w3c/web-platform-tests/css/css-cascade/scope-hover.html, imported/w3c/web-platform-tests/css/css-contain/contain-size-block-003.html, imported/w3c/web-platform-tests/css/css-contain/contain-size-inline-block-003.html, imported/w3c/web-platform-tests/css/css-view-transitions/massive-element-below-viewport-partially-onscreen-new.html, imported/w3c/web-platform-tests/css/css-view-transitions/massive-element-below-viewport-partially-onscreen-old.html, imported/w3c/web-platform-tests/css/selectors/invalidation/user-valid-user-invalid.html, imported/w3c/web-platform-tests/css/selectors/user-invalid.html, imported/w3c/web-platform-tests/css/selectors/valid-invalid-form-fieldset.html, imported/w3c/web-platform-tests/html/semantics/disabled-elements/disabled-event-dispatch.tentative.html, imported/w3c/web-platform-tests/html/semantics/invokers/invoketarget-button-event-dispatch.tentative.html, imported/w3c/web-platform-tests/html/semantics/selectors/pseudo-classes/readwrite-readonly.html, imported/w3c/web-platform-tests/mimesniff/media/media-sniff.window.html, imported/w3c/web-platform-tests/mixed-content/gen/worker-classic.http-rp/opt-in/fetch.https.html, imported/w3c/web-platform-tests/video-rvfc/request-video-frame-callback-webrtc.https.html, imported/w3c/web-platform-tests/webrtc-extensions/RTCRtpSynchronizationSource-captureTimestamp.html (failure)") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/34330 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/26182 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/27767 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/27665 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/32930 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/30589 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/6735 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/4873 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/30754 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/8480 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/37023 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/7339 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/7470 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/7963 "Passed tests") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/7277 "Built successfully") | | |

### CODE REVIEW

File: Source/cmake/OptionsGTK.cmake [hunk 1]
```diff
@@ -69,6 +68,8 @@ WEBKIT_OPTION_CONFLICT(USE_GTK4 USE_SOUP2)
 
 # Private options specific to the GTK port. Changing these options is
 # completely unsupported. They are intended for use only by WebKit developers.
+WEBKIT_OPTION_DEFINE(USE_LIBBACKTRACE "Whether to enable usage of libbacktrace." PRIVATE OFF)
```
Comment (aperezdc) [inline Source/cmake/OptionsGTK.cmake]: Why `PRIVATE`? This seems unrelated to the stated goal of the patch.

File: Source/cmake/OptionsWPE.cmake [hunk 2]
```diff
@@ -99,13 +99,13 @@ WEBKIT_OPTION_DEFINE(ENABLE_WPE_QT_API "Whether to enable support for the Qt5/QM
...
 WEBKIT_OPTION_DEFINE(USE_GBM "Whether to enable usage of GBM." PUBLIC ON)
 WEBKIT_OPTION_DEFINE(USE_LCMS "Whether to enable support for image color management using libcms2." PUBLIC ON)
-WEBKIT_OPTION_DEFINE(USE_LIBBACKTRACE "Whether to enable usage of libbacktrace." PUBLIC ON)
 WEBKIT_OPTION_DEFINE(USE_LIBDRM "Whether to enable usage of libdrm." PUBLIC ON)
 WEBKIT_OPTION_DEFINE(USE_OPENJPEG "Whether to enable support for JPEG2000 images." PUBLIC ON)
 WEBKIT_OPTION_DEFINE(USE_SOUP2 "Whether to enable usage of Soup 2 instead of Soup 3." PUBLIC OFF)
 WEBKIT_OPTION_DEFINE(USE_WOFF2 "Whether to enable support for WOFF2 Web Fonts." PUBLIC ON)
 
 # Private options specific to the WPE port.
+WEBKIT_OPTION_DEFINE(USE_LIBBACKTRACE "Whether to enable usage of libbacktrace." PRIVATE OFF)
```
Comment (aperezdc) [inline Source/cmake/OptionsWPE.cmake]: Ditto.

**Review by aperezdc — COMMENTED** [review]
Some distributions are applying the patch from https://github.com/ianlancetaylor/libbacktrace/pull/92 to add a `.pc` file to their libbacktrace packages. I would rather keep a few lines in a `.cmake` module file.

Also your commit log does not explain _why_ using pkg-config is an issue. Now, whether we want `ENABLE_LIBBACKTRACE=ON` by default or not, that's a different discussion.

File: Source/cmake/OptionsGTK.cmake (hunk 1 above)
Comment (mcatanzaro) [reply Source/cmake/OptionsGTK.cmake]: So my thinking here is that if there are no upstream releases yet, it's too early to expose the option to users. Having more PUBLIC options makes the feature list more confusing, so I would make it PUBLIC at the same time that we enable it by default.

**Comment by mcatanzaro** [comment]
> Some distributions are applying the patch from https://github.com/ianlancetaylor/libbacktrace/pull/92 to add a .pc file to their libbacktrace packages. I would rather keep a few lines in a .cmake module file.

What if upstream eventually merges a pkg-config with a different name, say `backtrace.pc`? It reminds me of the era when distros would maintain custom soversions for shared libraries when upstreams didn't version their libraries properly, which turned out to be a big mistake.

Shame there is no response from upstream in that issue. If Ian was willing to add the pkg-config file, then we wouldn't need to debate this....

**Comment by mcatanzaro** [comment]
It looks like this broke a couple API tests. I will figure out how to disable or skip them.

File: Source/cmake/OptionsGTK.cmake (hunk 1 above)
Comment (mcatanzaro) [reply Source/cmake/OptionsGTK.cmake]: OK?

**Comment by mcatanzaro** [comment]
So, I'm pretty sure we should not depend on a pkg-config file that does not exist upstream. Let's keep that.

Do you want me to split this into two pull requests, or is it OK to land all together like this?

File: Source/cmake/OptionsGTK.cmake (hunk 1 above)
Comment (aperezdc) [reply Source/cmake/OptionsGTK.cmake]: It's `PUBLIC` because we want the feature enabled by default, while allowing distributors to turn it off.

**Comment by aperezdc** [comment]
> So, I'm pretty sure we should not depend on a pkg-config file that does not exist upstream. Let's keep that.

Sure, we can re-add it later if/when upstream ships a `.pc` file.
 
> Do you want me to split this into two pull requests, or is it OK to land all together like this?

I would leave the option as `PUBLIC ON` and then we don't even need to consider wheter two different patches are needed 😉 

At the very, very least we want this to be `PUBLIC ${ENABLE_DEVELOPER_MODE}` because using libbacktrace gets us symbols in backtraces _even_ with symbol visibility attributes in use (and with LTO, and a few other cases).

**Comment by mcatanzaro** [comment]
I'll edit this pull request tomorrow to only remove the pkg-config stuff and not change the visibility or default value of the build option. (That said, it was me who originally encouraged this to be enabled by default, because I didn't look closely enough into libbacktrace and didn't realize it doesn't have any releases.)

> At the very, very least we want this to be PUBLIC ${ENABLE_DEVELOPER_MODE} because using libbacktrace gets us symbols in backtraces even with symbol visibility attributes in use (and with LTO, and a few other cases).

Honestly I have pretty low esteem for backtraces not generated by gdb, but I understand it's been difficult to add nicer core dump handling to the bots, and a libbacktrace backtrace is certainly a lot better than no backtrace. Still, this really seems primarily interesting for bots IMO.  Developers should basically always be using gdb.

**Comment by mcatanzaro** [comment]
> (That said, it was me who originally encouraged this to be enabled by default, because I didn't look closely enough into libbacktrace and didn't realize it doesn't have any releases.)

To be clear: this package doesn't exist in Debian or Fedora, and lacking releases I have no interest in packaging it. So I expect all distros will turn this OFF. I regret suggesting that we should default it to ON. Oh well; we can always change that later, if desired.

**Comment by aperezdc** [comment]
> I'll edit this pull request tomorrow to only remove the pkg-config stuff and not change the visibility or default value of the build option. (That said, it was me who originally encouraged this to be enabled by default, because I didn't look closely enough into libbacktrace and didn't realize it doesn't have any releases.)

Thanks, that sounds good.

> > At the very, very least we want this to be PUBLIC ${ENABLE_DEVELOPER_MODE} because using libbacktrace gets us symbols in backtraces even with symbol visibility attributes in use (and with LTO, and a few other cases).
> 
> Honestly I have pretty low esteem for backtraces not generated by gdb, but I understand it's been difficult to add nicer core dump handling to the bots, and a libbacktrace backtrace is certainly a lot better than no backtrace. Still, this really seems primarily interesting for bots IMO. Developers should basically always be using gdb.

OTOH, building in support for libbacktrace does help lots when using GDB to get a backtrace is not an option. Sometimes we don't have access to the hardware where issues arise, or it's underpowered to run full debug builds, and then having a backtrace with symbols and line numbers is much better than not having anything. It's also useful for release builds with symbols which get deployed to devices, although there we have a overlap with the Breakpad support.

**Comment by mcatanzaro** [comment]
> Thanks, that sounds good.

Oh, I already did this. It's ready for review again.

**Review by aperezdc — APPROVED** [review]
r=me, thanks @mcatanzaro
